diff --git a/libloragw/src/loragw_spi-stm32.c b/libloragw/src/loragw_spi-stm32.c
index a9ec563..6079921 100644
--- a/libloragw/src/loragw_spi-stm32.c
+++ b/libloragw/src/loragw_spi-stm32.c
@@ -158,8 +158,6 @@ int lgw_spi_open(void **spi_target_ptr) {
 
   speed_t ispeed = cfgetispeed(&options);
   speed_t ospeed = cfgetospeed(&options);
-  printf("baud rate in: 0%o\n", ispeed);
-  printf("baud rate out: 0%o\n", ospeed);
   
     *spi_device = dev;
     *spi_target_ptr = (void *)spi_device;
