diff --git a/libloragw/Makefile b/libloragw/Makefile
index 49b197e..d78a36c 100644
--- a/libloragw/Makefile
+++ b/libloragw/Makefile
@@ -10,7 +10,7 @@ CROSS_COMPILE ?=
 CC := $(CROSS_COMPILE)gcc
 AR := $(CROSS_COMPILE)ar
 
-CFLAGS := -O2 -Wall -Wextra -std=c99 -Iinc -I.
+CFLAGS := -O2 -Wall -Wextra -std=c99 -Iinc -I. -lm
 
 ### library.cfg configuration file processing
 
@@ -34,10 +34,11 @@ endif
 
 ### general build targets
 
-all: libloragw.a test_loragw_spi test_loragw_reg test_loragw_hal test_loragw_gps test_loragw_cal
+all: libloragw.a libloragw.so test_loragw_spi test_loragw_reg test_loragw_hal test_loragw_gps test_loragw_cal
 
 clean:
 	rm -f libloragw.a
+	rm -f libloragw.so
 	rm -f test_loragw_*
 	rm -f obj/*.o
 	rm -f inc/config.h
@@ -98,6 +99,9 @@ libloragw.a: obj/loragw_hal.o obj/loragw_gps.o obj/loragw_reg.o obj/loragw_spi.o
 endif
 	$(AR) rcs $@ $^
 
+libloragw.so: obj/loragw_hal.o obj/loragw_gps.o obj/loragw_reg.o obj/loragw_spi.o obj/loragw_aux.o
+	$(CC) -shared -lm -o $@ $^
+
 ### test programs
 
 test_loragw_spi: tst/test_loragw_spi.c libloragw.a
