--- a/libloragw/library.cfg	2016-04-20 23:42:00.000000000 +0200
+++ b/libloragw/library.cfg	2016-04-20 23:42:36.000000000 +0200
@@ -7,7 +7,10 @@
 #			      to ensure the right device will be opened on your platform.
 #	ftdi		FTDI SPI-over-USB bridge using libmpsse/libftdi/libusb
 
-CFG_SPI= native
+CFG_SPI= ftdi
+CFG_CHIP= sx1301
+CFG_RADIO= sx1257
+CFG_BAND= eu868
 
 ### Debug options ###
 # Set the DEBUG_* to 1 to activate debug mode in individual modules.
