--- a/libloragw/src/loragw_spi.ftdi.c	2016-04-21 16:32:14.484977349 +0200
+++ b/libloragw/src/loragw_spi.ftdi.c	2016-04-21 16:32:26.455856348 +0200
@@ -52,7 +52,7 @@
 
 /* parameters for a FT2232H */
 #define VID		0x0403
-#define PID		0x6010
+#define PID		0x6014
 
 /* -------------------------------------------------------------------------- */
 /* --- PUBLIC FUNCTIONS DEFINITION ------------------------------------------ */
