--- a/drivers/of/fdt.c
+++ b/drivers/of/fdt.c
@@ -956,6 +956,9 @@ int __init early_init_dt_scan_chosen(uns
 	p = of_get_flat_dt_prop(node, "bootargs", &l);
 	if (p != NULL && l > 0)
 		strlcpy(data, p, min((int)l, COMMAND_LINE_SIZE));
+	p = of_get_flat_dt_prop(node, "bootargs-append", &l);
+	if (p != NULL && l > 0)
+		strlcat(data, p, min_t(int, strlen(data) + (int)l, COMMAND_LINE_SIZE));
 
 	/*
 	 * CONFIG_CMDLINE is meant to be a default in case nothing else
--- a/arch/mips/kernel/setup.c
+++ b/arch/mips/kernel/setup.c
@@ -630,6 +630,7 @@ static void __init arch_mem_init(char **
 	struct memblock_region *reg;
 	extern void plat_mem_setup(void);
 
+	strlcpy(boot_command_line, arcs_cmdline, COMMAND_LINE_SIZE);
 	/* call board setup routine */
 	plat_mem_setup();
 
