--- a/configure.ac
+++ b/configure.ac
@@ -114,7 +114,7 @@ AC_ARG_ENABLE([libevent-regress],
 	[], [enable_libevent_regress=yes])
 AC_ARG_ENABLE([samples],
      AS_HELP_STRING([--disable-samples, skip building of sample programs]),
-	[], [enable_samples=yes])
+	[], [enable_samples=no])
 AC_ARG_ENABLE([function-sections],
      AS_HELP_STRING([--enable-function-sections, make static library allow smaller binaries with --gc-sections]),
 	[], [enable_function_sections=no])
@@ -791,10 +791,6 @@ fi
 
 # check if we have and should use openssl
 AM_CONDITIONAL(OPENSSL, [test "$enable_openssl" != "no" && test "$have_openssl" = "yes"])
-if test "x$enable_openssl" = "xyes"; then
-	AC_SEARCH_LIBS([ERR_remove_thread_state], [crypto eay32],
-		[AC_DEFINE(HAVE_ERR_REMOVE_THREAD_STATE, 1, [Define to 1 if you have ERR_remove_thread_stat().])])
-fi
 
 # Add some more warnings which we use in development but not in the
 # released versions.  (Some relevant gcc versions can't handle these.)
