diff --git a/lib/crypto.js b/lib/crypto.js
index d1c9eb5..d442777 100644
--- a/lib/crypto.js
+++ b/lib/crypto.js
@@ -234,6 +234,7 @@ Hash.prototype.update = function(data, encoding) {
   encoding = encoding || exports.DEFAULT_ENCODING;
   if (encoding === 'buffer' && typeof data === 'string')
     encoding = 'binary';
+  var data = new Buffer(data);
   this._binding.update(data, encoding);
   return this;
 };
