--- a/plugins/nfctype1.c	2016-06-17 18:49:00.615969207 +0200
+++ b/plugins/nfctype1.c	2016-06-17 18:49:25.024385011 +0200
@@ -29,6 +29,7 @@
 #include <stdbool.h>
 #include <string.h>
 #include <sys/socket.h>
+#include <sys/types.h>
 
 #include <linux/socket.h>
 
--- a/plugins/p2p.c	2016-06-17 18:57:26.720599462 +0200
+++ b/plugins/p2p.c	2016-06-17 18:57:02.508186934 +0200
@@ -30,6 +30,7 @@
 #include <string.h>
 #include <poll.h>
 #include <sys/socket.h>
+#include <sys/time.h>
 
 #include <linux/socket.h>
 
