--- a/util/Makefile.am
+++ b/util/Makefile.am
@@ -17,7 +17,7 @@ mount_fuse3_SOURCES = mount.fuse.c
 
 install-exec-hook:
 	-chmod u+s $(DESTDIR)$(bindir)/fusermount3
-	@if test ! -e $(DESTDIR)/dev/fuse; then \
+	-@if test ! -e $(DESTDIR)/dev/fuse; then \
 		$(MKDIR_P) $(DESTDIR)/dev; \
 		echo "mknod $(DESTDIR)/dev/fuse -m 0666 c 10 229 || true"; \
 		mknod $(DESTDIR)/dev/fuse -m 0666 c 10 229 || true; \
