--- a/drivers/mtd/devices/m25p80.c
+++ b/drivers/mtd/devices/m25p80.c
@@ -244,6 +244,12 @@ static int m25p_remove(struct spi_device
 {
 	struct m25p	*flash = spi_get_drvdata(spi);

+	// manfeel note: add spi flash reset code
+	flash->command[0] = 0x66;
+	spi_write(flash->spi, flash->command, 1);
+	flash->command[0] = 0x99;
+	spi_write(flash->spi, flash->command, 1);
+
 	/* Clean up MTD stuff. */
 	return mtd_device_unregister(&flash->mtd);
 }
@@ -318,6 +324,9 @@ static struct spi_driver m25p80_driver =
 	.id_table	= m25p_ids,
 	.probe	= m25p_probe,
 	.remove	= m25p_remove,
+
+	// manfeel, add shutdown method to reset spi flash
+	.shutdown = m25p_remove,

 	/* REVISIT: many of these chips have deep power-down modes, which
 	 * should clearly be entered on suspend() to minimize power use.
