--- a/arch/arm/boot/dts/qcom-ipq8064-c2600.dts
+++ b/arch/arm/boot/dts/qcom-ipq8064-c2600.dts
@@ -398,7 +398,6 @@
                 status {
 			label = "status:blue";
 			gpios = <&qcom_pinmux 53 0>;
-			default-state = "on";
 		};
                 ledgnr {
 			label = "ledgnr:blue";
--- /dev/null
+++ b/base-files/etc/diag.sh
@@ -0,0 +1,29 @@
+. /lib/functions/leds.sh
+. /lib/ipq806x.sh
+
+get_status_led() {
+	case $(ipq806x_board_name) in
+	c2600)
+		status_led="status:blue"
+		;;
+	esac
+}
+
+set_state() {
+	get_status_led
+
+	case "$1" in
+	preinit)
+		status_led_blink_preinit
+		;;
+	failsafe)
+		status_led_blink_failsafe
+		;;
+	preinit_regular)
+		status_led_blink_preinit_regular
+		;;
+	done)
+		status_led_on
+		;;
+	esac
+}
