--- a/squashfs-tools/mksquashfs.c	2017-09-24 16:56:56.167381381 -0400
+++ b/squashfs-tools/mksquashfs.c	2017-09-24 16:57:04.816572945 -0400
@@ -40,6 +40,7 @@
 #include <signal.h>
 #include <setjmp.h>
 #include <sys/mman.h>
+#include <sys/sysmacros.h>
 
 #ifndef linux
 #define __BYTE_ORDER BYTE_ORDER
--- a/squashfs-tools/unsquashfs.c	2017-09-24 17:06:40.593860859 -0400
+++ b/squashfs-tools/unsquashfs.c	2017-09-24 17:06:45.921873480 -0400
@@ -32,6 +32,7 @@
 #include <zlib.h>
 #include <sys/mman.h>
 #include <utime.h>
+#include <sys/sysmacros.h>
 
 #ifndef linux
 #define __BYTE_ORDER BYTE_ORDER
