--- a/Makefile	2014-04-12 18:59:48.378837000 +0800
+++ b/Makefile	2014-04-13 18:30:36.077405778 +0800
@@ -12,7 +12,7 @@
 
 EXTRA_CFLAGS += -I$(src)/include
 
-ccflags-y += -D__CHECK_ENDIAN__
+ccflags-y += -D__CHECK_ENDIAN__ -DCONFIG_88EU_AP_MODE
 
 CONFIG_AUTOCFG_CP = n
 
@@ -143,7 +143,7 @@
 all: modules
 
 modules:
-	$(MAKE) ARCH=$(ARCH) CROSS_COMPILE=$(CROSS_COMPILE) -C $(KSRC) M=$(shell pwd)  modules
+	$(MAKE) ARCH=$(ARCH) CROSS_COMPILE=$(CROSS_COMPILE) -C $(KSRC) M=$(PKG_BUILD_DIR)  modules
 
 strip:
 	$(CROSS_COMPILE)strip 8188eu.ko --strip-unneeded
