Index: linux-4.9.34/arch/arm/boot/dts/qcom-ipq8064-ap148.dts
===================================================================
--- linux-4.9.34.orig/arch/arm/boot/dts/qcom-ipq8064-ap148.dts
+++ linux-4.9.34/arch/arm/boot/dts/qcom-ipq8064-ap148.dts
@@ -225,66 +225,66 @@
 				ports {
 					#address-cells = <1>;
 					#size-cells = <0>;
-					port@0 {
+					cpu_port1: port@0 {
 						reg = <0>;
 						label = "cpu";
 						ethernet = <&gmac1>;
 						phy-mode = "rgmii";
 
- 						fixed-link {
- 							speed = <1000>;
- 							full-duplex;
- 						};
+						fixed-link {
+							speed = <1000>;
+							full-duplex;
+						};
 					};
 
 					port@1 {
 						reg = <1>;
 						label = "lan1";
+						cpu = <&cpu_port1>;
 						phy-handle = <&phy_port1>;
 					};
 
 					port@2 {
 						reg = <2>;
 						label = "lan2";
+						cpu = <&cpu_port1>;
 						phy-handle = <&phy_port2>;
 					};
 
 					port@3 {
 						reg = <3>;
 						label = "lan3";
+						cpu = <&cpu_port1>;
 						phy-handle = <&phy_port3>;
 					};
 
-					port@4 {
-						reg = <4>;
+					port@5{
+						reg = <5>;
 						label = "lan4";
-						phy-handle = <&phy_port4>;
+						cpu = <&cpu_port1>;
+						phy-handle = <&phy_port5>;
 					};
 
-					port@5 {
-						reg = <5>;
+					port@4 {
+						reg = <4>;
 						label = "wan";
-						phy-handle = <&phy_port5>;
+						cpu = <&cpu_port2>;
+						phy-handle = <&phy_port4>;
 					};
 
-					/*
-					 * Disabled until DSA supports multiple CPUs,
-					 * otherwise it causes undefined behavior.
-					 *
-					 * port@6 {
-					 *	reg = <6>;
-					 *	label = "cpu";
-					 *	ethernet = <&gmac2>;
-					 *	phy-mode = "sgmii";
-					 *
- 					 *	fixed-link {
- 					 *		speed = <1000>;
- 					 *		full-duplex;
- 					 *	};
-					 * };
-					 */
+					cpu_port2: port@6 {
+						reg = <6>;
+						label = "cpu";
+						ethernet = <&gmac2>;
+						phy-mode = "sgmii";
+
+						fixed-link {
+							speed = <1000>;
+							full-duplex;
+						};
+					};
 				};
-			};		
+			};
 		};
 
 		gmac1: ethernet@37200000 {
@@ -295,6 +295,8 @@
 			pinctrl-0 = <&rgmii2_pins>;
 			pinctrl-names = "default";
 
+			mac-address = [00 11 22 33 44 55];
+
 			fixed-link {
 				speed = <1000>;
 				full-duplex;
@@ -306,6 +308,8 @@
 			phy-mode = "sgmii";
 			qcom,id = <2>;
 
+			mac-address = [00 11 22 33 44 56];
+
 			fixed-link {
 				speed = <1000>;
 				full-duplex;
