--- a/autogen.sh	2016-09-15 09:17:23.000000000 +0200
+++ b/autogen.sh	2016-09-15 09:18:01.000000000 +0200
@@ -19,10 +19,10 @@
 # directory is set up as an SVN external.
 if test ! -e gtest; then
   echo "Google Test not present.  Fetching gtest-1.7.0 from the web..."
-  curl -O https://googletest.googlecode.com/files/gtest-1.7.0.zip
+  wget https://github.com/google/googletest/archive/release-1.7.0.zip -O gtest-1.7.0.zip
   unzip -q gtest-1.7.0.zip
   rm gtest-1.7.0.zip
-  mv gtest-1.7.0 gtest
+  mv googletest-release-1.7.0 gtest
 fi

 set -ex
