--- a/tests/fs-tests/integrity/integck.c
+++ b/tests/fs-tests/integrity/integck.c
@@ -31,7 +31,6 @@
 #include <getopt.h>
 #include <assert.h>
 #include <mntent.h>
-#include <execinfo.h>
 #include <bits/stdio_lim.h>
 #include <sys/mman.h>
 #include <sys/vfs.h>
@@ -262,8 +261,6 @@ static void check_failed(const char *con
 	 * 2. Compile with -rdynamic and -g gcc options
 	 * 3. Preferrably compile with -O0 to avoid inlining
 	 */
-	count = backtrace(addresses, 128);
-	backtrace_symbols_fd(addresses, count, fileno(stdout));
 	exit(EXIT_FAILURE);
 }
 
