--- a/libhandle/handle.c
+++ b/libhandle/handle.c
@@ -21,6 +21,7 @@
 #include "xfs.h"
 #include "handle.h"
 #include "parent.h"
+#include <linux/limits.h>
 
 /* just pick a value we know is more than big enough */
 #define	MAXHANSIZ	64
--- a/libhandle/jdm.c
+++ b/libhandle/jdm.c
@@ -21,6 +21,7 @@
 #include "handle.h"
 #include "jdm.h"
 #include "parent.h"
+#include <linux/limits.h>
 
 /* internal fshandle - typecast to a void for external use */
 #define FSHANDLE_SZ		8
