--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -84,6 +84,7 @@ dtb-$(CONFIG_ARCH_BCM_5301X) += \
 	bcm4709-netgear-r8000.dtb \
 	bcm4709-tplink-archer-c9-v1.dtb \
 	bcm47094-dlink-dir-885l.dtb \
+        bcm47094-linksys-ea9500.dtb \
 	bcm47094-luxul-xwr-3100.dtb \
 	bcm47094-netgear-r8500.dtb \
 	bcm94708.dtb \
--- /dev/null
+++ b/arch/arm/boot/dts/bcm47094-linksys-ea9500.dts
@@ -0,0 +1,149 @@
+/*
+ * Broadcom BCM47094 / BCM5301X ARM platform code.
+ * Basic DTS for Linksys EA9500
+ *
+ *
+ * Licensed under the GNU/GPL. See COPYING for details.
+ */
+
+/dts-v1/;
+
+#include "bcm47094.dtsi"
+#include "bcm5301x-nand-cs0-bch1.dtsi"
+
+/ {
+   compatible = "linksys,ea9500", "brcm,bcm47094", "brcm,bcm4708";
+   model = "Linksys EA9500";
+
+   chosen {
+       bootargs = "console=ttyS0,115200 earlyprintk";
+   };
+
+   memory {
+	reg = <0x00000000 0x08000000
+			0x88000000 0x08000000>;
+	};
+
+   leds {
+       compatible = "gpio-leds";
+
+       wps {
+           label = "bcm53xx:white:wps";
+           gpios = <&chipcommon 22 GPIO_ACTIVE_LOW>;
+           linux,default-trigger = "default-off";
+       };
+
+       usb2 {
+           label = "bcm53xx:green:usb2";
+           gpios = <&chipcommon 1 GPIO_ACTIVE_LOW>;
+           linux,default-trigger = "default-off";
+       };
+
+       usb3 {
+           label = "bcm53xx:green:usb3";
+           gpios = <&chipcommon 2 GPIO_ACTIVE_LOW>;
+           linux,default-trigger = "default-off";
+       };
+
+       power {
+           label = "bcm53xx:white:power";
+           gpios = <&chipcommon 4 GPIO_ACTIVE_HIGH>;
+           linux,default-trigger = "default-on";
+       };
+
+       wifibtn {
+           label = "bcm53xx:amber:wifibtn";
+           gpios = <&chipcommon 0 GPIO_ACTIVE_LOW>;
+           linux,default-trigger = "default-off";
+       };
+
+       wireless {
+           label = "bcm53xx:white:wireless";
+           gpios = <&chipcommon 5 GPIO_ACTIVE_HIGH>;
+           linux,default-trigger = "default-off";
+       };
+
+       bluebar1 {
+           label = "bcm53xx:white:bluebar1";
+           gpios = <&chipcommon 11 GPIO_ACTIVE_HIGH>;
+           linux,default-trigger = "default-off";
+       };
+
+       bluebar2 {
+           label = "bcm53xx:white:bluebar2";
+           gpios = <&chipcommon 12 GPIO_ACTIVE_HIGH>;
+           linux,default-trigger = "default-off";
+       };
+
+       bluebar3 {
+           label = "bcm53xx:white:bluebar3";
+           gpios = <&chipcommon 15 GPIO_ACTIVE_LOW>;	//actHigh -> actLow
+           linux,default-trigger = "default-off";
+       };
+
+       bluebar4 {
+           label = "bcm53xx:white:bluebar4";
+           gpios = <&chipcommon 18 GPIO_ACTIVE_HIGH>;
+           linux,default-trigger = "default-off";
+       };
+
+       bluebar5 {
+           label = "bcm53xx:white:bluebar5";
+           gpios = <&chipcommon 19 GPIO_ACTIVE_HIGH>;
+           linux,default-trigger = "default-off";
+       };
+
+       bluebar6 {
+           label = "bcm53xx:white:bluebar6";
+           gpios = <&chipcommon 20 GPIO_ACTIVE_HIGH>;
+           linux,default-trigger = "default-off";
+       };
+
+       bluebar7 {
+           label = "bcm53xx:white:bluebar7";
+           gpios = <&chipcommon 21 GPIO_ACTIVE_HIGH>;
+           linux,default-trigger = "default-off";
+       };
+
+       bluebar8 {
+           label = "bcm53xx:white:bluebar8";
+           gpios = <&chipcommon 8 GPIO_ACTIVE_HIGH>;
+           linux,default-trigger = "default-off";
+       };
+
+
+   };
+
+   gpio-keys {
+       compatible = "gpio-keys";
+       #address-cells = <1>;
+       #size-cells = <0>;
+
+       rfkill {
+           label = "WiFi";
+           linux,code = <KEY_RFKILL>;
+           gpios = <&chipcommon 16 GPIO_ACTIVE_LOW>;
+       };
+
+       wps {
+           label = "WPS";
+           linux,code = <KEY_WPS_BUTTON>;
+           gpios = <&chipcommon 3 GPIO_ACTIVE_LOW>;
+       };
+
+       reset {
+           label = "Reset";
+           linux,code = <KEY_RESTART>;
+           gpios = <&chipcommon 17 GPIO_ACTIVE_LOW>;
+       };
+
+   };
+};
+
+&usb2 {
+	vcc-gpio = <&chipcommon 13 GPIO_ACTIVE_HIGH>;
+};
+
+&usb3 {
+	vcc-gpio = <&chipcommon 14 GPIO_ACTIVE_HIGH>;
+};
